src/ALE/Recon1d_PPM_CW.F90

```diff
@@ -0,0 +1,418 @@
+  procedure :: dfdx => dfdx
+  !> Implementation of deallocation for PPM_CW
+  procedure :: destroy => destroy
+  !> Implementation of check reconstruction for the PPM_CW reconstruction
+  procedure :: check_reconstruction => check_reconstruction
+  !> Implementation of unit tests for the PPM_CW reconstruction
+  procedure :: unit_tests => unit_tests
+
+  !> Duplicate interface to init()
+  procedure :: init_parent => init
+  !> Duplicate interface to reconstruct()
+  procedure :: reconstruct_parent => reconstruct
+
+end type PPM_CW
+
+contains
+
+!> Initialize a 1D PPM_CW reconstruction for n cells
+subroutine init(this, n, h_neglect, check)
+  class(PPM_CW),     intent(out) :: this      !< This reconstruction
+  integer,           intent(in)  :: n         !< Number of cells in this column
+  real, optional,    intent(in)  :: h_neglect !< A negligibly small width used in cell reconstructions [H]
+  logical, optional, intent(in)  :: check     !< If true, enable some consistency checking
+
+  this%n = n
+
+  allocate( this%u_mean(n) )
+  allocate( this%ul(n) )
+  allocate( this%ur(n) )
+
+  ! This incurs an extra store of u_mean but by using PCM_CW
+  ! we avoid duplicating and testing more code
+  call this%PLM%init( n, h_neglect=h_neglect, check=check )
+
+  this%h_neglect = tiny( this%u_mean(1) )
+  if (present(h_neglect)) this%h_neglect = h_neglect
+  this%check = .false.
+  if (present(check)) this%check = check
+
+end subroutine init
+
+!> Calculate a 1D PPM_CW reconstructions based on h(:) and u(:)
+subroutine reconstruct(this, h, u)
+  class(PPM_CW), intent(inout) :: this !< This reconstruction
+  real,          intent(in)    :: h(*) !< Grid spacing (thickness) [typically H]
+  real,          intent(in)    :: u(*) !< Cell mean values [A]
+  ! Local variables
+  real :: h0, h1, h2, h3 ! Cell thickness h(k-2), h(k-1), h(k), h(k+1) in K loop [H]
+  real :: d12 ! h1 + h2  but used in the denominator so include h_neglect [H]
+  real :: h01_h112, h23_h122 ! Approximately 2/3 [nondim]
+  real :: ddh ! Approximately 0 [nondim]
+  real :: I_h12, I_h0123 ! Reciprocals of d12 and sum(h) [H-1]
+  real :: dul, dur ! Left and right cell PLM slopes [A]
+  real :: u0, u1, u2 ! Far left, left, and right cell values [A]
+  real :: edge ! Edge value between cell k-1 and k [A]
+  real :: u_min, u_max ! Minimum and maximum value across edge [A]
+  real :: a6 ! Colella and Woodward curvature [A]
+  real :: du ! Difference between edges across cell [A]
+  real :: slp(this%n) ! PLM slope [A]
+  integer :: k, n
+
+  n = this%n
+
+  ! First populate the PLM reconstructions
+  call this%PLM%reconstruct( h, u )
+  do k = 1, n
+    slp(k) = this%PLM%ur(k) - this%PLM%ul(k)
+  enddo
+  ! Extrapolate from interior for boundary PLM slopes
+  ! Note: this is not conventional but helps retain accuracy near top/bottom
+  ! boudaries and reduces the adverse influence of the boudnaries int he interior
```

Comment:
Here "boudnaries int he " should be "boundaries in the" .

Reply:
Fixed, thx